src/realm/db.cpp

```diff
@@ -624,6 +493,136 @@ DB::SharedInfo::SharedInfo(Durability dura, Replication::HistoryType ht, int hsv
+        m_info = m_reader_map.get_addr();
+        m_local_max_entry = m_info->readers.capacity();
+        REALM_ASSERT(sizeof(SharedInfo) + m_info->readers.compute_required_space(m_local_max_entry) == size);
+    }
+
+    void cleanup_versions(uint64_t& oldest_live_version, TopRefMap& top_refs, bool& any_new_unreachables)
+    {
+        std::lock_guard lock(m_mutex);
+        ensure_full_reader_mapping();
+        m_info->readers.purge_versions(oldest_live_version, top_refs, any_new_unreachables);
+    }
+
+    version_type get_newest_version()
+    {
+        return get_version_id_of_latest_snapshot().version;
+    }
+
+    VersionID get_version_id_of_latest_snapshot()
+    {
+        std::lock_guard lock(m_mutex);
+        ensure_full_reader_mapping();
+        uint_fast32_t index = m_info->readers.newest;
+        return {m_info->readers.get(index).version, index};
+    }
+
+    void release_read_lock(const ReadLockInfo& read_lock)
+    {
+        std::lock_guard lock(m_mutex);
+        // we should not need to call ensure_full_reader_mapping,
+        // since releasing a read lock means it has been grabbed
+        // earlier - and hence must reside in mapped memory:
+        REALM_ASSERT(read_lock.m_reader_idx < m_local_max_entry);
+        auto& r = m_info->readers.get(read_lock.m_reader_idx);
+        REALM_ASSERT(read_lock.m_version == r.version);
+        if (read_lock.m_is_frozen) {
+            --r.count_frozen;
+        }
+        else {
+            --r.count_live;
+        }
+    }
+
+    ReadLockInfo grab_read_lock(bool frozen, VersionID version_id = {})
+    {
+        ReadLockInfo read_lock;
+        std::lock_guard lock(m_mutex);
+        ensure_full_reader_mapping();
+        const bool pick_specific = version_id.version != VersionID().version;
+        auto newest = m_info->readers.newest.load();
+        REALM_ASSERT(newest != VersionList::nil);
+        read_lock.m_reader_idx = pick_specific ? version_id.index : newest;
+        bool picked_newest = read_lock.m_reader_idx == (unsigned)newest;
+        auto& r = m_info->readers.get(read_lock.m_reader_idx);
+        if (pick_specific && version_id.version != r.version)
+            throw BadVersion();
+        if (!picked_newest) {
+            if (frozen && r.count_frozen == 0 && r.count_live == 0)
+                throw BadVersion();
+            if (!frozen && r.count_live == 0)
+                throw BadVersion();
+        }
+        if (frozen) {
+            ++r.count_frozen;
+        }
+        else {
+            ++r.count_live;
+        }
+        read_lock.m_is_frozen = frozen;
+        read_lock.m_version = r.version;
+        read_lock.m_top_ref = static_cast<ref_type>(r.current_top);
+        read_lock.m_file_size = static_cast<size_t>(r.filesize);
+        return read_lock;
+    }
+
+    void add_version(ref_type new_top_ref, size_t new_file_size, uint64_t new_version)
+    {
+        std::lock_guard lock(m_mutex);
```

Comment:
Don't we need `ensure_full_reader_mapping()` here?

Reply:
Good catch. We didn't earlier, but with the new representation of the VersionList we do.